text/0000-at-least-one-feature.md

```diff
@@ -0,0 +1,183 @@
+
+   - Suppose it is on by default, or depended-upon by `bar`.
+     Oh no!
+     Existing `no-default-features = true` are now broken!
+     They want a feature set of `{bar}` which would correspond to the old `{}`, but there is no way to arrange it.
+
+In step two, we could "ret-con" the empty feature set with the `default` feature.
+But this is a trick that can only be pulled once.
+The second time around, we already have a default feature; we are out of luck.
+
+The previous attempts attempted to make new default features, or migrate the requested feature sets.
+But that is complex.
+There is exactly a simpler solution: simply require that *some* feature always be depended-upon.
+
+To see why this works, it helps to first see that step 2 was *already* broken.
+Here's the thing, even though there previously were not any features, that *doesn't* mean there were not any `no-default-features = true` users!
+Sure, it wouldn't do anything for crate with new features, but one can still use it.
+Then when just `bar` is added, we already have a problem, because the `default` feature will no "catch" all the existing users ---
+the mischievous users that were already using `no-default-features = true` will have their code broken!
+
+This brings us to the heart of the problem.
+So long as users are depending on "something", we can be careful to make sure those features keep their meaning.
+But when users are depending on nothing at all with `no-default-features = true` and an empty feature set, we have nothing to "hook into".
+The simple solution is just to rule out that problem entirely!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Packages with
+```toml
+[package]
+at-least-one-feature = true
+```
+are easier to maintain!
+You don't need to worry about the `no-default-features = true, features = []` case anymore.
+You can be sure that all consumers must dependent on either `default` or a regular named feature.
+
+Whenever you want to make existing functionality more conditional, simply extend the set of features the code relies on with a new feature, and ensure existing features depend on that feature.
+
+For example:
+```rust
+fn my_fun_that_allocates() { .. }
+
+#[cfg(all(feature = "foo", feature = "bar"))]
+fn my_weird_fun_that_allocates() { .. }
+```
+becomes:
+```rust
+#[cfg(feature = "baz")]
+fn my_fun_that_allocates() { .. }
+
+#[cfg(all(feature = "foo", feature = "bar", feature = "baz"))]
+fn my_weird_fun_that_allocates() { .. }
+```
+
+And the corresponding `Cargo.toml`:
+```toml
+[features]
+default = ["foo"]
+bar = ["foo"]
+```
+becomes:
+```toml
+[features]
+default = ["foo", "baz"]
+foo = ["baz"]
+bar = ["foo"] # no need to add "baz" because "bar" picks it up
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Depending on a
+```toml
+[package]
+at-least-one-feature = true
```

Comment:
It would be nice if this could provide a way for existing crates to move over to the new system without breaking compatibility. Perhaps add a `no-feature-selected = "feature"` key that allows to specify the single feature that is activated when no others are?

Reply:
I don't think it breaks compatibility. Because the solver knows to reject plans where the no-feature dep is mis-resovled, there are no issues. Not all downstream uses can upgrade, but all that can will.

In general I think it's good to think of version bumps as the *residual* of thing that the plan solver doesn't already know how to detect. Bumping a version bound is saying "sorry there is a problem I don't know how to directly say what it is, so I am going to insert this barrier instead".

Reply:
> I don't think it breaks compatibility. Because the solver knows to reject plans where the no-feature dep is mis-resovled, there are no issues. Not all downstream uses can upgrade, but all that can will.

I don't exactly understand your language here. Do you mean that cargo will not update a crate's dependency to a version with `at-least-one-feature = true`, if the depending (not dependent) crate does not specify a feature for the dependency?

Reply:
Exactly! Thanks for saying that more clearly :)

Reply:
Can you paraphrase what you mean by breaking? The most naive definitions "breaks my build" or "breaks my built executable" don't seem to apply here.

Reply:
This is exactly like is bumping the minimum Rust version a breaking change. Some people cannot upgrade (they are using an old Rust version) but no ones build breaks with incorrect uprades.

Reply:
I added material addressing this and making that comparison to minimum Rust versions.

Reply:
I think this needs more consideration than an immediate hand-waiving this as resolved. While never picking the version helps, I suspect there are still pitfalls. For example, another crate upgrading their version requirement will cause a resolver error despite "being compatible".

And I don't think MSRV is an appropriate precedence. That is dealing with the build environment while this is dealing with the API.

Reply:
I didn't mean to dismiss this, but just continue the conversation in a comment of what I wrote in the new RFC version. Otherwise anyone reading the RFC conversation will have no idea the issue is now discussed in the RFC, and I'll just have to inline what I write there in the comments too.

Reply:
Periodically trying to address concerns in the comments and then opening new threads I think is a good way to keep the conversation manual. Perhaps my "`at-least-one-feature = true` is not a breaking change" explanation is all wrong, but I think the first-order problem is the RFC originally neglected to discuss the issue at all. Now at is at least discussed, and so problems with that discussion (and I think it's better to poke holes in the RFC version than the version in the comments) are to me follow-up concerns.

All that said, if you still rather continue the conversation here that is fine.

Reply:
For me, the topic is breaking changes and spreading that out across multiple threads can make that harder to follow.

Reply:
OK Then, I'll keep it here.

> And I don't think MSRV is an appropriate precedence. That is dealing with the build environment while this is dealing with the API.

I went with `MSRV` because it is an example of something we used to not have, so it is nice to illustrate the point about "residual" breakage. But it not the only option. (Also once https://github.com/rust-lang/cargo/issues/9930 is added I don't think the "build environment" argument makes sense. Version bounds (only a lower bound because we assume there is no breaking changes in rustc) that effect version solving are bona fide dependencies, are they not?)

Is increasing a dependency bound by a narrow version a breaking change? I think the answer is also "no". Anyone stuck on the old version (private deps make this a bit harder to imagine, but public deps reveal it) will be unable to upgrade, but no breakage will occur.

Moreover, we can call this a breaking change, and that a breaking change, but does anyone benefit from bumping the version number? No. Those that already couldn't upgrade still can't. And those that previously *could* upgrade now also can't. "breaking" nomenclature aside, the practical difference is simply to rule at more perfectly good build plans.

-----

See what I wrote in https://github.com/rust-lang/rfcs/pull/3347#discussion_r1020777387. I think one thing worth taking from @tbu-'s RFC's is basically a one-off migration/ret-conning of `default-features = false`. Then we side-step that feature entirely. The restriction in this PR ensures that the "one off" approach is sufficient --- since going forward there won't be new `default-features = false` users we don't have to worry about that cropping up again.

Reply:
MSRV is a very different case that, just because we want to improve the experience for it, does not mean we take it as a license to do similar elsewhere.

>  but no breakage will occur.
> 
> does anyone benefit from bumping the version number

I would consider it a breaking change if a crate switched from `foo = "1.0"` to `foo = "1.0;<1.3"` which is effectively what we are talking about here.

This doesn't just affect "really old" crates where we can assume no one should upgrade any dependency.  People upgrade dependencies piece meal.  Some times they have a reason to hold back.

As I pointed out elsewhere, its not just a matter of gracefully being blocked from upgrade this dependency but being blocked from upgrading the dependency anywhere in the dependency tree.  I can't depend on an unrelated crate that needs a newer version of the crate that made the switch over.

Incorporating my idea elsewhere to encourage people to use this feature via `cargo new` defaults and edition defaults at least improves this by moving this away from a "only enabled once someone needs it" to "its enabled defensively".

Reply:
> Incorporating my idea elsewhere to encourage people to use this feature via cargo new defaults and edition defaults at least improves this by moving this away from a "only enabled once someone needs it" to "its enabled defensively".

Yeah I do like that part, to be clear.